### PR TITLE
#158286547 Add tests for dynamic filter

### PR DESCRIPTION
**Pull request by Victoradukwu**
#### What does this PR do?
This PR adds in tests for the the dynamic filter functionality

#### Description of Task to be completed?
* Add tests for the checks the functionalities of dynamic filter.

#### How should this be manually tested?
* Run `git fetch origin ch-dynamic-filter-tests-158286547:ch-dynamic-filter-tests-158286547`, then `git checkout ch-dynamic-filter-tests-158286547`
* Run the command `pytest tests/test_query_filter.py`, to see the test results.

#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?
[#158286547](https://www.pivotaltracker.com/story/show/158286547)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

**Comment by GbengaOshinaga** [comment]
@Victoradukwu Your `How should this be manually tested?` section does not follow the proper format we use, it should be `git fetch origin ch-dynamic-filter-tests-158286547:ch-dynamic-filter-tests-158286547`, then `git checkout ch-dynamic-filter-tests-158286547`

File: tests/test_query_filter.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class TestQueryFilter:
+    """
+    Class that holds methods for testing dynamic filter
+    """
+
+    def test_for_wrong_formatted_query(
+            self, client, auth_header):
+        """Chek the format of the filter string
```
Comment (GbengaOshinaga) [inline tests/test_query_filter.py]: Spelt `Check` wrong

Comment (Victoradukwu) [reply tests/test_query_filter.py]: Thank you, @GbengaOshinaga. Done.

File: tests/conftest.py [hunk 2]
```diff
@@ -163,3 +163,16 @@ def mock_request_obj_decoded_token():
...
+
+@pytest.fixture(scope="module")
+def multiple_categories():
+    """
+    Create the asset category records for testing
+    """
+    category_one = AssetCategory(name="Laptop")
+    category_two = AssetCategory(name="Chromebook")
+    category_three = AssetCategory(name="Apple")
+    category_one.save()
```
[the remark this answers is not in the document]
Comment (Victoradukwu) [reply tests/conftest.py]: @ibutiti  This has been done using the loop, but also takes advantage of the existing save() method of model operations.

File: tests/test_query_filter.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    def test_for_request_without_filter_query(
+            self, client, init_db, auth_header, multiple_categories):
+        """Returns all records when no filter string is provided
+
+        Tests if no filter query is specified. if no filter query
+        is specified, the search should returned all the records in the
+        specified category
+
+        Returns: dict: a dictionary of response with with status of success
```
[the remark this answers is not in the document]
Comment (Victoradukwu) [reply tests/test_query_filter.py]: The docstring comments have been taken care of.

File: tests/test_query_filter.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        """
+
+        response = client.get(
+            f'{api_v1_base_url}/asset-categories?where=name;eq,Laptop',
+            headers=auth_header)
+        response_json = json.loads(response.data.decode(CHARSET))
+
+        assert response.status_code == 400
+        assert response_json["status"] == "error"
+        assert response_json["message"] == "Invalid filter format:"\
```
[the remark this answers is not in the document]
Comment (Victoradukwu) [reply tests/test_query_filter.py]: Done

File: tests/conftest.py [hunk 5]
```diff
@@ -163,3 +164,24 @@ def mock_request_obj_decoded_token():
...
 
     return decoded_token
+
+
+@pytest.fixture(scope="module")
+def multiple_categories():
+    """
+    Create the asset category records for testing
+    """
+    categories = [
```
Comment (ibutiti) [inline tests/conftest.py]: A step in the right direction but there's still some repetition. How about you do it like this:
```
categories = ['Laptop', 'Chromebook', 'Apple']

for category in categories:
    db.session.add(AssetCategory(name=category))

db.session.commit()
```

File: tests/conftest.py [hunk 6]
```diff
@@ -163,3 +164,24 @@ def mock_request_obj_decoded_token():
...
+        AssetCategory(name="Laptop"),
+        AssetCategory(name="Chromebook"),
+        AssetCategory(name="Apple")
+    ]
+    for category in categories:
+        category.save()
```
Comment (ibutiti) [inline tests/conftest.py]: Yes, you're leveraging on the existing save method. However, if you look at the method closely, it's executing:
`db.session.add(self)` then `db.session.commit()` for each instance. We can make it more efficient by only executing a single commit by adding all instances in the session first then execute a single database write.

File: tests/test_query_filter.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        assert response.status_code == 200
+        assert response_json["status"] == "success"
+        assert len(response_json["data"]) == 3
+
+    def test_for_request_with_one_filter_query(
+            self, client, init_db, auth_header, multiple_categories):
+        """
+        Test a single filter condition
+        the reques should return only two records that satisfy the condition
```
Comment (ibutiti) [inline tests/test_query_filter.py]: Typo in `request`

File: tests/test_query_filter.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+    def test_invalid_filter_operator(self, dynamic_filter):
+        """
+        Assert that the filter_query method of DynamicFilter class raises
+        exception if the argument has a wrong operator. Valid operators
+        include: eq, ne, like, ge, gl, lt, gt
+        """
+        with pytest.raises(ValidationError):
+            dynamic_filter.filter_query(ImmutableMultiDict(
+                [('where', 'age,bad_comparator,19')]))
```
Comment (ibutiti) [inline tests/test_query_filter.py]: Nice. Please add 1 test for when the filter works as expected.

**Comment by brayoh** [comment]
@Victoradukwu Kindly fix the merge conflicts

**Comment by Victoradukwu** [comment]
The merge conflicts have been fixed, @brayoh

**Pull request merged**